### ISSUE TITLE
How to get the Best Model Weights (Across All Runs)

### ISSUE THREAD

**Issue by DoctorSlimm** [ultralytics/ultralytics#4096]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi All! Wondering how to extract the best model weights from all the /runs/detect/train{i} ... as the best model may not be the one in the last train{final_index} directory...? Or

**Comment by xaristeidou** [comment]
@DoctorSlimm When you train a model, in the end of training procedure there are to files exported in `/runs/detect/train/weights`. The first file is `last.pt` and the other file is `best.pt`. The last.pt has the weights from the last epoch, and the best.pt has the weights from the epoch which had the highest metrics (i think the highest mAP50-95).

**Comment by DoctorSlimm** [comment]
Thanks! Sometimes there is train1 train2 train3 etc. which all have their own best.pt and last.pt files etc... great to know that the best.pt is actually based on mAP specifically :)

**Comment by glenn-jocher** [comment]
Hi @DoctorSlimm,

That's correct! For each training run (for example, train1, train2, train3, etc.), a separate set of weights is produced and is stored in its own directory. Inside each directory, you'll find two files: `best.pt` and `last.pt`.

The `last.pt` file corresponds to the model's weights at the end of the final training epoch. On the other hand, the `best.pt` file contains the weights of the model from the epoch that achieved the highest overall validation performance, which is typically measured by mean Average Precision (mAP) across various Intersection over Union (IoU) thresholds.

In other words, if you're looking for the best weights across all training runs, you'll want to compare the metrics (like mAP) across the `best.pt` file from each run, not the `last.pt` files.

Hope that answers your question! Please feel free to reach out if further clarification is needed. 

Thanks, 
Glenn

**Comment by DoctorSlimm** [comment]
@glenn-jocher Thanks! I Love how you as the founder are so active answering everyones questions! One more thigns regarding metrics...

YOLOv8 Saves by Default the model with the best mAP metric or something like that, whereas for me I am most interested only in val/box_loss primarily, so the best weights saved are not the best for my metric of interest. 

TLDR: Is there a way to ensure the best.pt is the best for my metric of most interest (aside from saving all epoch checkpoints and reading from a graph)

Thanks! :)

**Comment by YK-1111** [comment]
Hi,
"the best.pt file contains the weights of the model from the epoch that achieved the highest overall validation performance, which is typically measured by mean Average Precision (mAP) across various Intersection over Union (IoU) thresholds."
Where can I find the precise method for determining the best .pt in YOLOv8?
Thank You

**Comment by YK-1111** [comment]
Hi, may i still ask question here? (please let me know if this is not allow)
"the best.pt file contains the weights of the model from the epoch that achieved the highest overall validation performance, which is typically measured by mean Average Precision (mAP) across various Intersection over Union (IoU) thresholds."
Where can I find the precise method for calculating and determining the best .pt in YOLOv8?
Thank You

**Comment by glenn-jocher** [comment]
@YK-1111 hi there,

Absolutely, you are welcome to ask questions here, and we're here to help!

In YOLOv8, the `best.pt` file is determined by monitoring a specific metric during the validation phase after each training epoch. By default, this metric is the mean Average Precision (mAP) at different IoU thresholds, as it provides a balanced view of the model's precision and recall across all classes and object sizes.

The calculation of mAP involves predicting bounding boxes on the validation set and comparing these with the ground truth annotations. For each class, precision-recall curves are generated based on the model predictions at varying confidence thresholds. The area under these curves is averaged over all classes and IoU thresholds to produce the final mAP score.

During training, after each epoch, the model is evaluated on the validation set, and the mAP is calculated. The weights from the epoch which achieve the highest mAP score are saved as `best.pt`. This means that the model saved in `best.pt` is the one that had the best balance of precision and recall on your validation dataset, according to the mAP metric.

Please note that if you're interested in prioritizing a different metric for performance evaluation, such as validation box loss, this can often be configured in the training script or via the command line arguments when initiating your training run. This will allow you to save the `best.pt` according to your preferred metric.

Hope this clarifies the process for determining the `best.pt` weights in YOLOv8. If you have more questions or need further assistance, feel free to reach out.

Best regards,
Glenn Jocher

**Comment by sarahsftri** [comment]
is there any way to look the mAP per epoch as the one in training for the validation?

**Comment by glenn-jocher** [comment]
@sarahsftri yes, mAP per epoch for validation is typically logged during training. You can review this information in the training logs or visualize it using TensorBoard, where mAP and other metrics are plotted for each epoch. Refer to the Ultralytics Docs for instructions on monitoring training progress.

**Comment by sarahsftri** [comment]
> @sarahsftri yes, mAP per epoch for validation is typically logged during training. You can review this information in the training logs or visualize it using TensorBoard, where mAP and other metrics are plotted for each epoch. Refer to the Ultralytics Docs for instructions on monitoring training progress.

ClearML Task: created new task id=15041257fc88422fbbc512aed0e78e3c
2023-12-15 00:17:20,059 - clearml.Repository Detection - WARNING - Could not read Jupyter Notebook: No module named 'nbconvert'
2023-12-15 00:17:20,108 - clearml.Task - INFO - Storing jupyter notebook directly as code
ClearML results page: https://app.clear.ml/projects/3246a72475664b7ca4de2cc5d19783c9/experiments/15041257fc88422fbbc512aed0e78e3c/output/log
New https://pypi.org/project/ultralytics/8.0.227 available  Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.0.200  Python-3.11.6 torch-2.1.0+cu121 CUDA:0 (NVIDIA GeForce RTX 3050 Laptop GPU, 4096MiB)
engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=YOLO_DATA/data.yaml, epochs=2, patience=50, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train7, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, vid_stride=1, stream_buffer=False, line_width=None, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, tracker=botsort.yaml, save_dir=runs\detect\train7
Overriding model.yaml nc=80 with nc=1

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]                  
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]                 
 22        [15, 18, 21]  1    751507  ultralytics.nn.modules.head.Detect           [1, [64, 128, 256]]           
Model summary: 225 layers, 3011043 parameters, 3011027 gradients, 8.2 GFLOPs

Transferred 319/355 items from pretrained weights
TensorBoard: Start with 'tensorboard --logdir runs\detect\train7', view at http://localhost:6006/
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed 
train: Scanning [D:\SKRIPSI\YOLO_NEW\YOLO_DATA\train\labels.cache...](file:///D:/SKRIPSI/YOLO_NEW/YOLO_DATA/train/labels.cache...) 7115 images, 6999 backgrounds, 0 corrupt: 100%|██████████| 8131/8131 [00:00<?, ?it/s]
val: Scanning [D:\SKRIPSI\YOLO_NEW\YOLO_DATA\valid\labels.cache...](file:///D:/SKRIPSI/YOLO_NEW/YOLO_DATA/valid/labels.cache...) 4017 images, 3459 backgrounds, 0 corrupt: 100%|██████████| 4017/4017 [00:00<?, ?it/s]
Plotting labels to runs\detect\train7\labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.002, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to runs\detect\train7
Starting training for 2 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        1/2      2.22G      1.459      8.176      1.202          0        640: 100%|██████████| 509/509 [02:05<00:00,  4.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 126/126 [00:39<00:00,  3.21it/s]
                   all       4017        559      0.732      0.683      0.742      0.451

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        2/2      2.13G      1.424      2.307      1.236          2        640: 100%|██████████| 509/509 [01:56<00:00,  4.38it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 126/126 [00:28<00:00,  4.38it/s]
                   all       4017        559      0.693      0.758      0.774      0.494

2 epochs completed in 0.087 hours.
Optimizer stripped from runs\detect\train7\weights\last.pt, 6.2MB
Optimizer stripped from runs\detect\train7\weights\best.pt, 6.2MB

Validating runs\detect\train7\weights\best.pt...
Ultralytics YOLOv8.0.200  Python-3.11.6 torch-2.1.0+cu121 CUDA:0 (NVIDIA GeForce RTX 3050 Laptop GPU, 4096MiB)
Model summary (fused): 168 layers, 3005843 parameters, 0 gradients, 8.1 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 126/126 [00:27<00:00,  4.62it/s]
                   all       4017        559      0.693       0.76      0.774      0.494
Speed: 0.2ms preprocess, 2.0ms inference, 0.0ms loss, 1.4ms postprocess per image
Results saved to runs\detect\train7
2023-12-15 00:26:14,091 - clearml.storage - INFO - Starting upload: runs\detect\train7\weights\best.pt => https://files.clear.ml/project_plate/plate_yolo.15041257fc88422fbbc512aed0e78e3c/models/best.pt
2023-12-15 00:26:40,550 - clearml.Task - INFO - Completed model upload to https://files.clear.ml/project_plate/plate_yolo.15041257fc88422fbbc512aed0e78e3c/models/best.pt


in train mAP that show is from train or val?

**Comment by glenn-jocher** [comment]
@sarahsftri the mAP reported during training corresponds to the performance on the validation data, not the training data. This metric reflects how well the model is likely to perform on unseen data. Keep an eye on the validation mAP to track your model's progress and avoid overfitting. 📈😊

**Comment by sarahsftri** [comment]
just to be clear yolov8 only provides mAP for the validation data only right?

**Comment by sarahsftri** [comment]
> just to be clear yolov8 only provides mAP for the validation data only right?

or is there any way so i can look the mAP for the train data?

**Comment by glenn-jocher** [comment]
@sarahsftri yes, that's correct. YOLOv8 provides the mAP for validation data only, as this is standard practice to evaluate model generalization. mAP for training data is typically not calculated to keep training efficient and because it's less indicative of model performance on unseen data. 🎯✅
